docs/decisions/0035-embedding-abstraction.md

```diff
@@ -0,0 +1,257 @@
+
+```csharp
+public class EmbeddingContent<TEmbedding> : KernelContent where TEmbedding : unmanaged
+{
+    public EmbeddingContent(
+        IReadOnlyList<ReadOnlyMemory<TEmbedding>> data,
+        string? modelId = null,
+        object? innerContent = null,
+        IReadOnlyDictionary<string, object?>? metadata = null)
+        : base(innerContent, modelId, metadata)
+    {
+        this.Data = data;
+    }
+
+    public IReadOnlyList<ReadOnlyMemory<TEmbedding>> Data { get; set; }
+}
+
+public interface IEmbeddingGenerationService : IAIService
+{
+    Task<IReadOnlyList<EmbeddingContent<double>>> GenerateEmbeddingsAsync(
+        IList<string> data,
+        Kernel? kernel = null,
+        PromptExecutionSettings? executionSettings = null,
+        CancellationToken cancellationToken = default);
+    
+    Task<IReadOnlyList<EmbeddingContent<double>>> GenerateEmbeddingsAsync(
+        IList<ImageContent> data,
+        Kernel? kernel = null,
+        PromptExecutionSettings? executionSettings = null,
+        CancellationToken cancellationToken = default);
+}
+```
+
+Pros:
+- Allows for generating embeddings from different types of data.
+- Allows for parameterizing the query.
+- Allows for returning metadata.
+
+Cons:
+- No generic interface.
+- One common interface for different types of data. Many models don't support all types of data.
+
+### Option 4 [Proposed] - Same as option 2 or 3 but with non-generic EmbeddingContent
+
+Similar to option 2 or 3, but with a non-generic `EmbeddingContent` class.
+
+This option forces the conversion of embeddings data always to `double`.
+
+```csharp
+public class EmbeddingContent : KernelContent
+{
+    public EmbeddingContent(
+        IReadOnlyList<ReadOnlyMemory<float>> data,
+        string? modelId = null,
+        object? innerContent = null,
+        IReadOnlyDictionary<string, object?>? metadata = null)
+        : base(innerContent, modelId, metadata)
+    {
+        this.Data = data;
+    }
+
+    public IReadOnlyList<ReadOnlyMemory<double>> Data { get; set; }
+}
+```
+
+Pros:
+- Simplicity.
+
+Cons:
+- We cannot return embeddings with different types.
+
+### Option 5 [Proposed] - One interface for embeddings generation with metadata and query parameters
+
+Similar to option 3 but with only one metod with KernelContent param.
+
+```csharp
+public class EmbeddingContent<TEmbedding> : KernelContent where TEmbedding : unmanaged
+{
+    public EmbeddingContent(
+        IReadOnlyList<ReadOnlyMemory<TEmbedding>> data,
+        string? modelId = null,
+        object? innerContent = null,
+        IReadOnlyDictionary<string, object?>? metadata = null)
+        : base(innerContent, modelId, metadata)
+    {
+        this.Data = data;
+    }
+
+    public IReadOnlyList<ReadOnlyMemory<TEmbedding>> Data { get; set; }
+}
+
+public interface IEmbeddingGenerationService : IAIService
+{
+    Task<IReadOnlyList<EmbeddingContent<double>>> GenerateEmbeddingsAsync(
```

Comment:
If we will have one interface for embeddings, that will return `EmbeddingContent<double>`, then we don't need to make `EmbeddingContent` as generic, because only one `double` type will be used. So, it looks like Option 4 and Option 5 can be merged together, please correct me if I'm wrong.